### PR TITLE
Add gslb pool algorithm as a configurable GS knob

### PR DESCRIPTION
**Pull request by hemantavi**
This commit introduces the GslbService Pool Algorithm as a configurable
knob as part of the GDP and the GSLB Host Rule objects. To use this knob,
one has to set the required field:

```
poolAlgorithmSettings:
  lbAlgorithm:
  hashMask:
  geoFallback:
    lbAlgorithm:
    hashMask:
```

For `lbAlgorithm`, the possible values are:
1. `GSLB_ALGORITHM_ROUND_ROBIN`
2. `GSLB_ALGORITHM_TOPOLOGY`
3. `GSLB_ALGORITHM_CONSISTENT_HASH` (`hashMask` field must be used for
this case).
4. `GSLB_ALGORITHM_GEO` (`geoFallback` field must be used for this case).

For `GSLB_ALGORITHM_GEO`, a `geoFallback` algorithm must be provided.
The `lbAlgorithm` for `geoFallback` can only have either `GSLB_ALGORITHM_CONSISTENT_HASH`
or `GSLB_ALGORITHM_ROUND_ROBIN`.

Ingestion layer changes:
- Added the field as part of the global filter and the gslb host rule
  objects.
- Any changes to the algorithm properties will prompt a push for all the
  objects to layer 2 (without any exceptions).
- Regular priority override for GSLB Host Rule's properties apply to
  algorithms too (in which case, GDP's algorithm will be ignored and
  GSLB Host Rule's algorithm will be considered.)
- Added a checksum function to calculate the checksum for the
 `poolAlgorithmSettings`.

Graph layer changes:
- Any add/update/delete for an object will prompt an update of the GS.
- Checksum updates and object copy functions now include algorithm as
  well.

Rest layer changes:
- Evaluate the requried alogrithm settings to be set (derived from the
  GS graph).
- Parse the algorithm settings in the controller_obj_cache.go file.

Integration tests:
- Added tests for transitions between valid and invalid combinations
  for both GDP and GSLB Host Rule objects.

### CODE REVIEW

**Comment by sudswasavi** [comment]
```poolAlgorithmSettings:
  lbAlgorithm:
  hashMask:
  geoFallback:
    lbAlgorithm:
    hashMask:```

When do you use the hashMask outside? 
The struct on first glance appears a bit confusing.

File: go.sum [hunk 1]
```diff
@@ -1135,6 +1136,7 @@ k8s.io/code-generator v0.18.6/go.mod h1:TgNEVx9hCyPGpdtCWA34olQYLkh3ok9ar7XfSsr8
 k8s.io/code-generator v0.19.0-beta.2/go.mod h1:6NiFnMML+4VaG+kHKew+dc+SBC3Q+5NTNQIHdJJbG3w=
 k8s.io/code-generator v0.19.2 h1:7uaWJll6fyCPj2j3sfNN1AiY2gZU1VFN2dFR2uoxGWI=
 k8s.io/code-generator v0.19.2/go.mod h1:moqLn7w0t9cMs4+5CQyxnfA/HV8MF6aAVENF+WZZhgk=
```
Comment (sudswasavi) [inline go.sum:1138]: you you need both versions of the code generator? In fact there are multiple.

File: go.sum [hunk 2]
```diff
@@ -1151,6 +1153,7 @@ k8s.io/gengo v0.0.0-20200428234225-8167cfdcfc14 h1:t4L10Qfx/p7ASH3gXCdIUtPbbIueg
 k8s.io/gengo v0.0.0-20200428234225-8167cfdcfc14/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
 k8s.io/gengo v0.0.0-20200728071708-7794989d0000 h1:XgICMZutMLbopSVIJJrhUun6Hbuh1NTZBv2sd0lvypU=
 k8s.io/gengo v0.0.0-20200728071708-7794989d0000/go.mod h1:aG2eeomYfcUw8sE3fa7YdkjgnGtyY56TjZlaJJ0ZoWo=
+k8s.io/gengo v0.0.0-20201113003025-83324d819ded h1:JApXBKYyB7l9xx+DK7/+mFjC7A9Bt5A93FPvFD0HIFE=
```
Comment (sudswasavi) [inline go.sum:1156]: same here - maybe a go mod tidy would be good to run.

**Comment by hemantavi** [comment]
@sudswasavi 
Sure, let me explain what the struct means:
```
  lbAlgorithm:
  hashMask:
  geoFallback:
    lbAlgorithm:
    hashMask:
```
The outer `lbAlgorithm` can have any one of 4 values:
- GSLB_ALGORITHM_ROUND_ROBIN
- GSLB_ALGORITHM_TOPOLOGY
- GSLB_ALGORITHM_CONSISTENT_HASH
- GSLB_ALGORITHM_GEO

The outer `hashMask` is only used in case of `GSLB_ALGORITHM_CONSISTENT_HASH`.
For `lbAlgorithm` as `GSLB_ALGORITHM_GEO`, we need a `geoFallback` setting provided. This field again has two sub-fields: `lbAlgorithm` and `hashMask`. This inner `lbAlgorithm` can only have one of the two values:
- GSLB_ALGORITHM_ROUND_ROBIN
- GSLB_ALGORITHM_CONSISTENT_HASH

For the innger `lbAlgorithm` as `GSLB_ALGORITHM_CONSISTENT_HASH`, the innger `hashMask` must be used.

File: gslb/test/integration/third_party_vips/int_test.go [hunk 3]
```diff
@@ -175,8 +175,8 @@ func SyncFromTestIngestionLayer(key string, wg *sync.WaitGroup) error {
 }
 
 func SyncFromTestNodesLayer(key string, wg *sync.WaitGroup) error {
-	gslbutils.Logf("recived key from graph layer: %s", key)
-	graphKeyChan <- key
+	gslbutils.Logf("recieved key from graph layer: %s", key)
+	// graphKeyChan <- key
```
Comment (sudswasavi) [inline gslb/test/integration/third_party_vips/int_test.go:179]: can you pls remove this?

Comment (hemantavi) [reply gslb/test/integration/third_party_vips/int_test.go]: Removed. Thanks.

File: go.sum (hunk 1 above)
Comment (hemantavi) [reply go.sum:1149]: I don't, but for some reason, running `go mod tidy` adds more lines for `k8s.io/apiextensions-apiserver`.

**Review by sudswasavi — APPROVED** [review]

**Review by chauhanshubham — APPROVED** [review]

**Pull request merged**